channeldb/meta.go

```diff
@@ -80,3 +88,92 @@ func putDbVersion(metaBucket kvdb.RwBucket, meta *Meta) error {
 	byteOrder.PutUint32(scratch, meta.DbVersionNumber)
 	return metaBucket.Put(dbVersionKey, scratch)
 }
+
+// OptionalMeta structure holds the database optional migration information.
+type OptionalMeta struct {
+	// Versions is a set that contains the versions that have been applied.
+	// When saved to disk, only the indexes are stored.
+	Versions map[uint64]string
+}
+
+func (om *OptionalMeta) String() string {
+	s := ""
+	for index, name := range om.Versions {
+		s += fmt.Sprintf("%d: %s", index, name)
+	}
+	if s == "" {
+		s = "empty"
+	}
+	return s
+}
+
+// fetchOptionalMeta reads the optional meta from the database.
+func (d *DB) fetchOptionalMeta() (*OptionalMeta, error) {
+	om := &OptionalMeta{
+		Versions: make(map[uint64]string),
+	}
+
+	err := kvdb.View(d, func(tx kvdb.RTx) error {
+		metaBucket := tx.ReadBucket(metaBucket)
+		if metaBucket == nil {
+			return ErrMetaNotFound
+		}
+
+		vBytes := metaBucket.Get(optionalVersionKey)
+		// Exit early if nothing found.
+		if vBytes == nil {
+			return nil
+		}
+
+		// Read the versions' length.
+		r := bytes.NewReader(vBytes)
+		vLen, err := tlv.ReadVarInt(r, &[8]byte{})
+		if err != nil {
+			return err
+		}
+
+		// Write the version index.
+		for i := uint64(0); i < vLen; i++ {
+			version, err := tlv.ReadVarInt(r, &[8]byte{})
+			if err != nil {
+				return err
+			}
+			om.Versions[version] = optionalVersions[i].name
+		}
+
+		return nil
+	}, func() {})
```

Comment:
Even though we don't know of people running nodes on etcd if they do, it's safer if we reset any external state properly, like in this current block we update the `om.Versions` map which is external to the closure.